llama-index-core/llama_index/core/workflow/result.py

```diff
@@ -0,0 +1,13 @@
+from typing import Any
+
+from llama_index.core.workflow.context import Context
+
+
+class WorkflowResult:
```

Comment:
I would implement this as an `asyncio.Future` so we can keep using `run` as usual. Additionally, we can add a new method to workflow like `create_task` (that was already on the todo list when we discussed how to simplify the stream consuming) to expose the streaming and stepwise execution. So this would be:
```python
import asyncio


class WorkflowHandler(asyncio.Task):
    async def stream_events(self):
        for i in range(10):
            await asyncio.sleep(0)
            yield f"Streaming event {i}"


class WorkflowResult(asyncio.Future):
    def __init__(self, *args, ctx=None, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self.ctx = ctx

    def __str__(self) -> str:
        return str(self.result())


class Workflow:
    def create_task(self, coro):
        loop = asyncio.get_event_loop()
        return WorkflowHandler(coro, loop=loop)

    async def run(self, ctx=None) -> WorkflowResult:
        res = WorkflowResult(ctx=ctx)
        await asyncio.sleep(1)  # simulate the workflow running
        res.set_result("Whatever the stop event contained")
        return res


async def main():
    # Run workflow in background and stream events (advanced)
    w = Workflow()
    r = w.create_task(w.run())
    async for e in r.stream_events():
        print(e)

    print(await r)

    # Run workflow in foreground and just wait for the result, same as before (basic)
    w = Workflow()
    print(await w.run())


asyncio.run(main())
```

We can find better names for `create_task` and `WorkflowHandler` to make it more obvious what they do in the context of a Workflow. I think by having `WorkflowHandler` we finally solve the problem of managing multiple sessions, while `WorkflowResult` will make the run method future proof, possibly avoiding more breaking changes in the future.